ExpressLRS/ExpressLRS
# Unified ESP targets

# Problem
OMG! So many targets and firmwares!

# Solution
Unified ESP targets for ESP32 TX modules and ESP8285 RXes.

This is a single unified firmware for ESP targets that loads a hardware descriptor (`hardware.json`) that is read from the SPIFFS partition or from the end of the firmware on startup in that order.
If the `hardware.json` file is not found the module boots straight to wifi mode and you can go to the `/hardware.html` page where there is a drop target for dropping pre-defined hardware files or you can manually enter the values for your specific target.

Also, the user-defines are also included in a SPIFFS file (`options.json`) or at the end of the firmware and can be changed/setup by going to `http://elrs-tx.local/options.html`. The binding phrase to UID script is included on the web-page for simplicity as well.

Also, the "device name" as sent to the LUA script is also stored in a SPIFFS file called `device.ini` and at the end of the firmware. This is a file with a single line with the text string for the device name.

The 3 files can be directly downloaded from the device by going to the appropriate URL e.g. http://elrs-tx.local/hardware.json. Extra functionality has been added so these files can be uploaded using HTTP PUT to the same URL. Also added a `/reboot` which reboots the device after you've uploaded the 3 control files (useful for Configurator).

When compiling in the traditional way, there is a post build task that attaches to the json files to the end of the firmware, so it should be indistinguishable from previous firmwares.

This also solves the problem of the proliferation of many user-defined DIY targets where there are only slight differences. see #1405.
Now a user can easily create their own target and keep a copy of their hardware file and upload it again if they do a full chip erase.

# How to test supported targets
1. "Build and flash" one of the supported ESP based targets (or other upload method)
2. 🎉 

# How to test generic targets
1. Use build and flash `Unified_2400_TX_via_UART` or other unified target.
2. goto the wifi page http://elrs_tx.local and you should see the target configuration page
3. drag and drop one of the target files from the `hardware` directory and press the `Save target configuration` button at the bottom of the page.
4. once the file is saved a popup asking you to reboot is presented and you should reboot the device
6. 🎉 

# What has been tested (TX)
- Axis Thor
- BETAFPV 900 Micro
- BETAFPV 2400 Micro
- BETAFPV 2400 Micro 1W
- DIY E28
- DIY DupleTX
- DIY TTGO 900 V2
- HappyModel ES24 Pro
- iFlight 2400
- iFlight 900
- Jumper AION Nano
- Jumper T-Pro
- Namimno Flash OLED
- Radiomaster Zorro
- Vantac Lite

# What has been tested (RX)
- BETAFPV Nano 2400
- EMAX 2400
- Generic 900
- Generic 2400
- Generic 2400 PWMP 5, 6, 7
- iFlight 2400
- Namimno FLASH
- Namimno FLASH PA
- Vantac 2400
